File: back-end/app/services/risk_service.py
from datetime import timedelta, datetime
import logging
from typing import List, Optional, Tuple, Dict, Any

from sqlalchemy import select, desc

from coclib.extensions import db
from coclib.models import PlayerSnapshot
from .snapshot_service import get_clan as refresh_clan

_WEIGHTS = {
    "war": 0.40,  # 40 pts when clan is actively warring
    "idle": 0.35,  # ↑ from 0.25 → emphasise recency of idleness
    "don_deficit": 0.15,
    "don_drop": 0.10,
}
_MAX = 100
_WAR_ATTACKS_TOTAL = 2
_DEFICIT_CEIL = 0.50
_DROP_CEIL = 0.30
_CLAN_WAR_WINDOW = 42

logger = logging.getLogger(__name__)


# Idle buckets → percentage contribution on the idle axis
def _idle_pct_from_days(days: int) -> float:
    """Return the idle axis percentage given whole *days* since last activity."""
    if days >= 4:
        return 1.0  # sitting duck
    if days == 3:
        return 0.75  # getting intense
    if days == 2:
        return 0.50  # mild concern
    return 0.0  # active within 24‑48h → no idle penalty


def _clamp01(x: float) -> float:
    return max(0.0, min(1.0, x))


def _latest_war_snapshot(history: List[PlayerSnapshot]) -> Optional[PlayerSnapshot]:
    """Newest snapshot that still contains war info (player was in the roster)."""
    for snap in reversed(history):          # newest → oldest
        if snap.war_attacks_used is not None:
            return snap
    return None

def _infer_attack_cap(history: List[PlayerSnapshot]) -> int:
    """Best-guess the max attacks available in the observed war."""
    cap = max(
        (snap.war_attacks_used or 0)
        for snap in history
        if snap.war_attacks_used is not None
    )
    return cap or _WAR_ATTACKS_TOTAL


def _clan_has_war(history_map: dict[str, list]) -> bool:
    """Return *True* if ANY member has war attacks in the recent window."""
    cutoff = history_map["now"] - timedelta(days=_CLAN_WAR_WINDOW) # noqa
    for snapshots in history_map["all"]:
        if any(s.ts >= cutoff and s.war_attacks_used is not None for s in snapshots):
            return True
    return False


async def get_history(player_tag: str, days: int = 30):
    """Return **oldest→newest** snapshots for *player_tag* limited to *days*."""
    cutoff = db.func.now() - timedelta(days=days)
    stmt = (
        select(PlayerSnapshot)
        .where(PlayerSnapshot.player_tag == player_tag, PlayerSnapshot.ts >= cutoff) # noqa
        .order_by(desc(PlayerSnapshot.ts))
    )
    res = db.session.execute(stmt).scalars().all()
    return list(reversed(res))


def _axes_data(
    history: List[PlayerSnapshot], clan_history_map: dict | None = None
) -> Tuple[Dict[str, Any], datetime]:
    if not history:
        return {
            "war": 0.0,
            "idle": 0.0,
            "deficit": 0.0,
            "drop": 0.0,
            "war_used": None,
            "war_cap": None,
            "idle_days": 0,
        }, datetime.utcnow()

    latest = history[-1]
    if len(history) < 2:
        prev = history[0]
    else:
        prev = history[-8] if len(history) >= 8 else history[0]

    war_snap = _latest_war_snapshot(history)
    if war_snap is None:
        war_miss_pct = 0.0
        war_used = None
        cap = None
    else:
        cap = _infer_attack_cap(history)
        war_used = war_snap.war_attacks_used
        war_miss_pct = _clamp01((cap - war_used) / cap)
        if (
            war_used == 0
            and datetime.utcnow() - war_snap.ts < timedelta(hours=24)
        ):
            # War likely still in progress – don't penalize yet
            war_miss_pct = 0.0

    clan_active = True
    if clan_history_map is not None:
        clan_active = _clan_has_war(clan_history_map)
    if not clan_active:
        war_miss_pct = 0.0

    last_change = next(
        (
            s
            for s in reversed(history)
            if s.trophies != latest.trophies or s.donations != latest.donations
        ),
        history[0],
    )
    activity_ts = latest.last_seen or last_change.ts
    idle_days = (datetime.utcnow() - activity_ts).days
    idle_pct = _idle_pct_from_days(idle_days)

    ratio = latest.donations / max(latest.donations_received, 1)
    deficit_pct = _clamp01((_DEFICIT_CEIL - ratio) / _DEFICIT_CEIL)

    drop_ratio = (prev.donations - latest.donations) / max(prev.donations, 1)
    drop_pct = _clamp01(drop_ratio / _DROP_CEIL)

    return (
        {
            "war": war_miss_pct,
            "idle": idle_pct,
            "deficit": deficit_pct,
            "drop": drop_pct,
            "war_used": war_used,
            "war_cap": cap,
            "idle_days": idle_days,
        },
        activity_ts,
    )


def score(
    history: List[PlayerSnapshot], clan_history_map: dict | None = None
) -> tuple[int, datetime]:
    """Compute the 0-100 risk score for a single member."""
    axes, activity_ts = _axes_data(history, clan_history_map)

    raw = (
        _WEIGHTS["war"] * axes["war"]
        + _WEIGHTS["idle"] * axes["idle"]
        + _WEIGHTS["don_deficit"] * axes["deficit"]
        + _WEIGHTS["don_drop"] * axes["drop"]
    )

    return int(round(raw * _MAX)), activity_ts


def score_breakdown(
    history: List[PlayerSnapshot], clan_history_map: dict | None = None
) -> Tuple[int, datetime, List[Dict[str, Any]]]:
    """Return risk score along with per-axis point breakdown."""
    axes, activity_ts = _axes_data(history, clan_history_map)

    war_pts = int(round(_WEIGHTS["war"] * axes["war"] * _MAX))
    idle_pts = int(round(_WEIGHTS["idle"] * axes["idle"] * _MAX))
    deficit_pts = int(round(_WEIGHTS["don_deficit"] * axes["deficit"] * _MAX))
    drop_pts = int(round(_WEIGHTS["don_drop"] * axes["drop"] * _MAX))

    breakdown: List[Dict[str, Any]] = []
    if war_pts:
        if axes["war_used"] is None:
            reason = "not in war roster"
        elif axes["war_used"] == 0:
            reason = "no war attacks used"
        else:
            missed = (axes["war_cap"] or 0) - axes["war_used"]
            reason = f"missed {missed} war attack{'s' if missed != 1 else ''}"
        breakdown.append({"points": war_pts, "reason": reason})
    if idle_pts:
        reason = f"inactive for {axes['idle_days']} day{'s' if axes['idle_days'] != 1 else ''}"
        breakdown.append({"points": idle_pts, "reason": reason})
    if deficit_pts:
        breakdown.append({"points": deficit_pts, "reason": "donation deficit"})
    if drop_pts:
        breakdown.append({"points": drop_pts, "reason": "donations dropped"})

    total = war_pts + idle_pts + deficit_pts + drop_pts
    return total, activity_ts, breakdown


async def clan_at_risk(clan_tag: str) -> list[dict]:
    """Return sorted risk breakdown for every member in *clan_tag*.

    The latest clan snapshot is fetched first so member data stays fresh.
    """

    await refresh_clan(clan_tag.upper())

    # Get the latest snapshot for each member, then compute history & score
    subq = (
        select(
            PlayerSnapshot.player_tag,
            db.func.max(PlayerSnapshot.ts).label("max_ts"),
        )
        .where(PlayerSnapshot.clan_tag == clan_tag)
        .group_by(PlayerSnapshot.player_tag)
        .subquery()
    )
    latest_rows = select(PlayerSnapshot).join(
        subq,
        (PlayerSnapshot.player_tag == subq.c.player_tag)
        & (PlayerSnapshot.ts == subq.c.max_ts),
    )
    players = db.session.execute(latest_rows).scalars().all()

    results = []
    for p in players:
        hist = await get_history(p.player_tag, 30)
        score_val, last_seen_ts, breakdown = score_breakdown(hist)
        iso_val = last_seen_ts.isoformat()
        logger.debug("last_seen raw for %s: %s", p.player_tag, iso_val)
        results.append(
            {
                "player_tag": p.player_tag,
                "name": p.name,
                "risk_score": score_val,
                "last_seen": iso_val,
                "risk_breakdown": breakdown,
            }
        )

    return sorted(results, key=lambda r: r["risk_score"], reverse=True)

File: clash-data/src/main/java/com/clanboards/clashdata/service/RiskService.java
package com.clanboards.clashdata.service;

import com.clanboards.clashdata.entity.PlayerSnapshot;
import java.time.Clock;
import java.time.LocalDateTime;
import java.time.temporal.ChronoUnit;
import java.util.*;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

@Service
public class RiskService {

  private static final Logger logger = LoggerFactory.getLogger(RiskService.class);

  // Risk calculation constants
  private static final Map<String, Double> DEFAULT_WEIGHTS =
      Map.of(
          "war", 0.40,
          "idle", 0.35,
          "don_deficit", 0.15,
          "don_drop", 0.10);
  private static final int MAX_SCORE = 100;
  private static final int WAR_ATTACKS_TOTAL = 2;
  private static final double DEFICIT_CEIL = 0.50;
  private static final double DROP_CEIL = 0.30;
  private static final int CLAN_WAR_WINDOW = 42;

  private final PlayerSnapshotService playerSnapshotService;
  private final SnapshotService snapshotService;
  private final Clock clock;

  @Autowired
  public RiskService(
      PlayerSnapshotService playerSnapshotService, SnapshotService snapshotService, Clock clock) {
    this.playerSnapshotService = playerSnapshotService;
    this.snapshotService = snapshotService;
    this.clock = clock;
  }

  public List<Map<String, Object>> getClanAtRisk(String clanTag, Map<String, Double> weights) {
    logger.debug("Getting at-risk members for clan: {}", clanTag);

    // Refresh clan snapshot to ensure membership list is current
    snapshotService.getClan(clanTag);

    // Get latest snapshots for all clan members
    List<PlayerSnapshot> latestSnapshots = playerSnapshotService.getLatestSnapshotsForClan(clanTag);

    List<Map<String, Object>> results = new ArrayList<>();

    for (PlayerSnapshot snapshot : latestSnapshots) {
      List<PlayerSnapshot> history =
          playerSnapshotService.getPlayerHistory(snapshot.getPlayerTag(), 30);

      if (history.isEmpty()) {
        continue;
      }

      RiskCalculationResult result = calculateRiskScoreWithBreakdown(history, null, weights);

      Map<String, Object> playerRisk = new HashMap<>();
      playerRisk.put("player_tag", snapshot.getPlayerTag());
      playerRisk.put("name", snapshot.getName());
      playerRisk.put("risk_score", result.score);
      playerRisk.put("last_seen", result.lastSeen.toString() + "Z");
      playerRisk.put("risk_breakdown", result.breakdown);

      results.add(playerRisk);
    }

    // Sort by risk score descending
    results.sort(
        (a, b) -> Integer.compare((Integer) b.get("risk_score"), (Integer) a.get("risk_score")));

    logger.debug("Found {} at-risk members for clan: {}", results.size(), clanTag);
    return results;
  }

  public int calculateRiskScore(
      List<PlayerSnapshot> history,
      Map<String, List<PlayerSnapshot>> clanHistoryMap,
      Map<String, Double> weights) {
    RiskCalculationResult result =
        calculateRiskScoreWithBreakdown(history, clanHistoryMap, weights);
    return result.score;
  }

  private RiskCalculationResult calculateRiskScoreWithBreakdown(
      List<PlayerSnapshot> history,
      Map<String, List<PlayerSnapshot>> clanHistoryMap,
      Map<String, Double> weights) {
    if (history.isEmpty()) {
      return new RiskCalculationResult(0, LocalDateTime.now(clock), new ArrayList<>());
    }

    PlayerSnapshot latest = history.get(history.size() - 1);
    PlayerSnapshot prev =
        history.size() < 2
            ? history.get(0)
            : (history.size() >= 8 ? history.get(history.size() - 8) : history.get(0));

    // Calculate war risk
    WarRiskData warRisk = calculateWarRisk(history, clanHistoryMap);

    // Calculate idle risk
    IdleRiskData idleRisk = calculateIdleRisk(history, latest);

    // Calculate donation deficit risk
    double deficitRisk = calculateDeficitRisk(latest);

    // Calculate donation drop risk
    double dropRisk = calculateDropRisk(latest, prev);

    // Merge weights
    Map<String, Double> actualWeights = mergeWeights(weights);

    // Calculate total score
    double rawScore =
        actualWeights.get("war") * warRisk.percentage
            + actualWeights.get("idle") * idleRisk.percentage
            + actualWeights.get("don_deficit") * deficitRisk
            + actualWeights.get("don_drop") * dropRisk;

    int totalScore = (int) Math.round(rawScore * MAX_SCORE);

    // Build breakdown
    List<Map<String, Object>> breakdown =
        buildRiskBreakdown(actualWeights, warRisk, idleRisk, deficitRisk, dropRisk);

    return new RiskCalculationResult(totalScore, idleRisk.lastActivityTime, breakdown);
  }

  private WarRiskData calculateWarRisk(
      List<PlayerSnapshot> history, Map<String, List<PlayerSnapshot>> clanHistoryMap) {
    Optional<PlayerSnapshot> latestWarSnap =
        history.stream()
            .filter(s -> s.getWarAttacksUsed() != null)
            .max(
                Comparator.comparing((PlayerSnapshot s) -> s.getWarAttacksUsed())
                    .thenComparing(PlayerSnapshot::getTs));

    if (latestWarSnap.isEmpty()) {
      return new WarRiskData(0.0, null, null, "not in war roster");
    }

    PlayerSnapshot warSnap = latestWarSnap.get();
    int warUsed = warSnap.getWarAttacksUsed();
    int cap =
        Math.max(
            history.stream()
                .filter(s -> s.getWarAttacksUsed() != null)
                .mapToInt(PlayerSnapshot::getWarAttacksUsed)
                .max()
                .orElse(0),
            WAR_ATTACKS_TOTAL);

    double warMissPct = Math.max(0.0, Math.min(1.0, (double) (cap - warUsed) / cap));

    // Don't penalize if war is still in progress (less than 24 hours ago and 0 attacks)
    if (warUsed == 0 && ChronoUnit.HOURS.between(warSnap.getTs(), LocalDateTime.now(clock)) < 24) {
      warMissPct = 0.0;
    }

    // Check if clan is actively warring (simplified - assume active for now)
    boolean clanActive = true; // Simplified implementation

    if (!clanActive) {
      warMissPct = 0.0;
    }

    String reason;
    if (warUsed == 0) {
      reason = "no war attacks used";
    } else {
      int missed = cap - warUsed;
      reason = String.format("missed %d war attack%s", missed, missed != 1 ? "s" : "");
    }

    return new WarRiskData(warMissPct, warUsed, cap, reason);
  }

  private IdleRiskData calculateIdleRisk(List<PlayerSnapshot> history, PlayerSnapshot latest) {
    // Find last activity (last change in trophies or donations, or last_seen)
    LocalDateTime lastActivity =
        latest.getLastSeen() != null ? latest.getLastSeen() : latest.getTs();

    for (int i = history.size() - 1; i >= 0; i--) {
      PlayerSnapshot current = history.get(i);
      if (i == 0
          || !Objects.equals(current.getTrophies(), latest.getTrophies())
          || !Objects.equals(current.getDonations(), latest.getDonations())) {
        if (current.getLastSeen() != null && current.getLastSeen().isAfter(lastActivity)) {
          lastActivity = current.getLastSeen();
        }
        break;
      }
    }

    long idleDays = ChronoUnit.DAYS.between(lastActivity, LocalDateTime.now(clock));
    double idlePct = getIdlePercentageFromDays((int) idleDays);

    return new IdleRiskData(idlePct, (int) idleDays, lastActivity);
  }

  private double getIdlePercentageFromDays(int days) {
    if (days >= 4) return 1.0;
    if (days == 3) return 0.75;
    if (days == 2) return 0.50;
    return 0.0;
  }

  private double calculateDeficitRisk(PlayerSnapshot latest) {
    double ratio = (double) latest.getDonations() / Math.max(latest.getDonationsReceived(), 1);
    return Math.max(0.0, Math.min(1.0, (DEFICIT_CEIL - ratio) / DEFICIT_CEIL));
  }

  private double calculateDropRisk(PlayerSnapshot latest, PlayerSnapshot prev) {
    double dropRatio =
        (double) (prev.getDonations() - latest.getDonations()) / Math.max(prev.getDonations(), 1);
    return Math.max(0.0, Math.min(1.0, dropRatio / DROP_CEIL));
  }

  private Map<String, Double> mergeWeights(Map<String, Double> customWeights) {
    Map<String, Double> weights = new HashMap<>(DEFAULT_WEIGHTS);
    if (customWeights != null) {
      for (Map.Entry<String, Double> entry : customWeights.entrySet()) {
        if (weights.containsKey(entry.getKey()) && entry.getValue() != null) {
          weights.put(entry.getKey(), entry.getValue());
        }
      }
    }
    return weights;
  }

  private List<Map<String, Object>> buildRiskBreakdown(
      Map<String, Double> weights,
      WarRiskData warRisk,
      IdleRiskData idleRisk,
      double deficitRisk,
      double dropRisk) {
    List<Map<String, Object>> breakdown = new ArrayList<>();

    int warPts = (int) Math.round(weights.get("war") * warRisk.percentage * MAX_SCORE);
    int idlePts = (int) Math.round(weights.get("idle") * idleRisk.percentage * MAX_SCORE);
    int deficitPts = (int) Math.round(weights.get("don_deficit") * deficitRisk * MAX_SCORE);
    int dropPts = (int) Math.round(weights.get("don_drop") * dropRisk * MAX_SCORE);

    if (warPts > 0) {
      Map<String, Object> warBreakdown = new HashMap<>();
      warBreakdown.put("points", warPts);
      warBreakdown.put("reason", warRisk.reason);
      breakdown.add(warBreakdown);
    }

    if (idlePts > 0) {
      Map<String, Object> idleBreakdown = new HashMap<>();
      idleBreakdown.put("points", idlePts);
      idleBreakdown.put(
          "reason",
          String.format("inactive for %d day%s", idleRisk.days, idleRisk.days != 1 ? "s" : ""));
      breakdown.add(idleBreakdown);
    }

    if (deficitPts > 0) {
      Map<String, Object> deficitBreakdown = new HashMap<>();
      deficitBreakdown.put("points", deficitPts);
      deficitBreakdown.put("reason", "donation deficit");
      breakdown.add(deficitBreakdown);
    }

    if (dropPts > 0) {
      Map<String, Object> dropBreakdown = new HashMap<>();
      dropBreakdown.put("points", dropPts);
      dropBreakdown.put("reason", "donations dropped");
      breakdown.add(dropBreakdown);
    }

    return breakdown;
  }

  // Helper classes for risk calculation data
  private static class RiskCalculationResult {
    final int score;
    final LocalDateTime lastSeen;
    final List<Map<String, Object>> breakdown;

    RiskCalculationResult(int score, LocalDateTime lastSeen, List<Map<String, Object>> breakdown) {
      this.score = score;
      this.lastSeen = lastSeen;
      this.breakdown = breakdown;
    }
  }

  private static class WarRiskData {
    final double percentage;
    final Integer used;
    final Integer cap;
    final String reason;

    WarRiskData(double percentage, Integer used, Integer cap, String reason) {
      this.percentage = percentage;
      this.used = used;
      this.cap = cap;
      this.reason = reason;
    }
  }

  private static class IdleRiskData {
    final double percentage;
    final int days;
    final LocalDateTime lastActivityTime;

    IdleRiskData(double percentage, int days, LocalDateTime lastActivityTime) {
      this.percentage = percentage;
      this.days = days;
      this.lastActivityTime = lastActivityTime;
    }
  }
}
